### ISSUE TITLE
Large merges take long time to complete (>60 seconds)

### ISSUE THREAD

**Issue by itaiad200** [treeverse/lakeFS#2690]
Reported on Slack.

Few hundred thousands objects merge fails after 30 seconds with (UI):
```
time="2021-11-10T15:33:02Z" level=trace msg="HTTP call ended" func=pkg/httputil.TracingMiddleware.func1.1 file="build/pkg/httputil/tracing.go:149" host=xxxx<http:///> method=POST path=/api/v1/repositories/aif-xxxx-xxxx/refs/xxxx-xxxx-processing-a6c0f/merge/main request_body="[123 125]" request_id=da585880-e027-41bb-b3cc-b64bfa08a818 response_body="{\"message\":\"merge in CommitManager: apply ns=s3://lakefs-data/aif-xxxx-xxxx<s3://lakefs-data/xxx> id=3fe239d80f74e" response_headers="map[Content-Type:[application/json] X-Request-Id:[da585880-e027-41bb-b3cc-b64bfa08a818]]" sent_bytes=0 service_name=rest_api status_code=500 took=30.008900206s
```

lakectl:
```
Command

/home/lakefs $ lakectl merge lakefs://aif-xxx-xxx/xxx-xxx-processing-a6c0f lakefs://aif-xxx-xxx/main -c /tmp/.lakectl.yaml --l
og-level trace
DEBU[0000]/build/cmd/lakectl/cmd/root.go:67 github.com/treeverse/lakefs/cmd/lakectl/cmd.glob..func66() loaded configuration from file                fields.file=/tmp/.lakectl.yaml file=/tmp/.lakectl.yaml
Source: lakefs://aif-xxx-xxx/xxx-xxx-processing-a6c0f
Destination: lakefs://aif-xxx-xxx/main
504 Gateway Timeout
```
Logs from Pod:
```
time="2021-11-10T18:19:39Z" level=trace msg="HTTP call ended" func=pkg/httputil.TracingMiddleware.func1.1 file="build/pkg/httputil/tracing.go:149" host=lakefs.ai.us.lmco.com method=POST path=/api/v1/repositories/aif-xxx-xxx/refs/xxx-xxx-processing-a6c0f/merge/main request_body="[123 125]" request_id=fc0348ba-0c85-4674-921f-253dd6839b66 response_body="{\"message\":\"merge in CommitManager: apply ns=s3://lakefs-data/aif-xxx-xxx id=3fe239d80f74e" response_headers="map[Content-Type:[application/json] X-Request-Id:[fc0348ba-0c85-4674-921f-253dd6839b66]]" sent_bytes=0 service_name=rest_api stat us_code=500 took=29.992150768s
```

**Comment by johnnyaug** [comment]
I would guess that there is a configured timeout in a network component on the way to lakeFS, like a load balancer. 
The error message and the status code are a result of the context being cancelled after the connection was closed.

**Comment by itaiad200** [comment]
That was my initial suspicion as well, but it reproduced when using lakectl on the same instance where lakeFS was running.

**Comment by tvanderwal-lmco** [comment]
I had an issue in my lakectl config file where it was still routing through dns. After updating it to point to `localhost:8000` it was able to complete the commit.

**Comment by tvanderwal-lmco** [comment]
OK I think I found the problem and this can probably be closed. 

In case anyone comes across this same issue and is using OpenShift, there is a default timeout of 30s added to routes and can be configured with an annotation. https://docs.openshift.com/container-platform/4.9/networking/routes/route-configuration.html

**Comment by itaiad200** [comment]
Thanks @tvanderwal-lmco. Now that we verified that the timeouts are not triggered by lakeFS, we still need to investigate why large merges are taking too long. Changed the issue title to reflect the current issue.

**Comment by itaiad200** [comment]
@guy-har I can see how #2710 reduces the amount of operations (commit/merge) that takes more than 60 seconds, but does not prevents it entirely. If our approach is that operations on millions of objects may take more than that, I think we should also explore other directions too, like making those operations async.

**Comment by arielshaqed** [comment]
> @guy-har I can see how #2710 reduces the amount of operations (commit/merge) that takes more than 60 seconds, but does not prevents it entirely. If our approach is that operations on millions of objects may take more than that, I think we should also explore other directions too, like making those operations async.

Obviously if we have no choice then we should make it async.  However I would prefer to make almost any effort to speed things up and avoid making anything async.  Async operations make ops extremely difficult, because in almost all cases you still need to determine success or failure.

Yes it may add complexity to the code, but that complexity is in one place and therefore manageable.  OTOH async operations spread immediately to everything that they touch.  Suddenly users need to poll, or use a (reliable!) queue, or otherwise handle this complexity.  Plus if we make it async but keep it long-running -- it will still take too long to run.

I believe that we can and should speed things up further for common cases.

**Comment by guy-har** [comment]
@itaiad200, It doesn't prevent entirely. It optimized only by 50%. By implementing the [Next step](https://github.com/treeverse/lakeFS/issues/2710#issuecomment-969394066) It should prevent entirely the cases we investigated. You could create cases that the optimization won't work for them. 
I agree that we should explore other directions like making those operations async.
It shouldn't be one direction or the other, both directions could work together (i.e optimizing by "fast-forwarding" ranges where we can, and where we can't, write the ranges asynchronously ).  Anyway exploring other directions is a great idea.

**Comment by guy-har** [comment]
Just to sum up the current status:

## The issues
- Large merges are taking a long time (> 60 seconds)
- Users with configured timeout in a network component on the way to lakeFS (such as load balancer) could not merge

## In progress

- Prevent timeout with keep-alive (research + implementation ~ 1 week) - #2796
- Optimize merge by fast-forwarding ranges ( resolved in #2710 and reduced merge time by 50% )
- Optimize merge by fast-forwarding for consecutive ranges - #2795 (~ 1 month)

## Next steps
- In case that timeout could not be prevented using keep-alive, consider adding async operations
- Investigate in more possible optimizations
  - Parallelization of commit/merge
  -  Techniques to write to memory and flush to disk
  - Optimize Pebble cache for large merges


For better understanding, read [here](https://docs.lakefs.io/understand/data-model.html) about our data model

**Comment by guy-har** [comment]
Merge now takes less than a minute in all the cases we are aware of. Closing this issue for now.
